src/core/util/data_version.pl

```diff
@@ -0,0 +1,180 @@
+
+:- begin_tests(read_data_version_header_tests).
+
+test("empty header", [fail]) :-
+    read_data_version_header_('', _).
+
+test("empty label and value", [fail]) :-
+    read_data_version_header_(':', _).
+
+test("empty label", [fail]) :-
+    read_data_version_header_(':value', _).
+
+test("empty value", [fail]) :-
+    read_data_version_header_('label:', _).
+
+test("extraneous colons", [fail]) :-
+    read_data_version_header_('label:value1:value2', _).
+
+test("short label", [fail]) :-
+    read_data_version_header_('labe:value', _).
+
+test("short value", [fail]) :-
+    read_data_version_header_('label:valu', _).
+
+test("pass") :-
+    read_data_version_header_('label:value', data_version(label, value)).
+
+:- end_tests(read_data_version_header_tests).
+
+write_data_version_header(data_version(Data_Version_Label, Data_Version_Value)) :-
+    atom(Data_Version_Label),
+    atom(Data_Version_Value),
+    !,
+    format("TerminusDB-Data-Version: ~s:~s~n", [Data_Version_Label, Data_Version_Value]).
+write_data_version_header(no_data_version) :-
+    !.
+write_data_version_header(Data_Version) :-
+    throw(error(unexpected_argument_instantiation(write_data_version_header, Data_Version), _)).
+
+/**
+ * transaction_data_version(+Object, -Data_Version) is semidet.
+ *
+ * Look for the data version of a transaction/validation object. Fail if Object
+ * does not have a data version.
+ */
+transaction_data_version(Object, Data_Version) :-
+    utils:do_or_die(
+        (   is_dict(Object),
+            (   transaction_object{} :< Object
+            ;   validation_object{} :< Object
+            )),
+        error(unexpected_argument_instantiation('transaction_data_version/2', Object), _)),
+    % Check if Object has a parent repository object and set that.
+    (   get_dict(parent, Object, Parent)
+    ->  Repo_Object = Parent
+    ;   Repo_Object = Object
+    ),
+    transaction_data_version_(Object.descriptor, Repo_Object, Data_Version).
+
+/**
+ * transaction_data_version(+Object, +Objects, -Data_Version) is semidet.
+ *
+ * Look for the data version of a transaction/validation object using Objects to
+ * look for Object's repository transaction/validation object. Fail if Object
+ * does not have a data version.
+ */
+transaction_data_version(Object, Objects, Data_Version) :-
+    utils:do_or_die(
+        (   is_dict(Object),
+            is_list(Objects),
+            (   transaction_object{} :< Object
+            ;   validation_object{} :< Object
+            )),
+        error(unexpected_argument_instantiation('transaction_data_version/3', objects(Object, Objects)), _)),
+    Descriptor = Object.descriptor,
+    (   transaction_data_version_(Descriptor, Object, Data_Version)
+    ->  true
+    ;   get_dict(repository_descriptor, Descriptor, Repo_Descriptor),
+        % Search Objects for the matching repository object.
+        member(Repo_Object, Objects),
+        get_dict(descriptor, Repo_Object, Repo_Descriptor),
```

Comment:
after we found a match we should cut, ensuring that no choice point is left.

Reply:
Nice one!

Reply:
https://github.com/terminusdb/terminusdb/pull/911/commits/d8bc2413a5d6e5d4800afbf1322d3e60239459fa